In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import scipy.stats as stats
import statsmodels.api as sm
import statsmodels.formula.api as smf
from statsmodels.formula.api import ols
import math
import statsmodels.api as sm
import statsmodels.formula.api as smf
from statsmodels.formula.api import ols

In [3]:
df = pd.read_csv('problem6.txt',sep= '\s+',header=None) # read the data set
df.columns = (['Y','X1','X2'])
df.head() 

Y    X1   X2
0  0.0  32.0  3.0
1  0.0  45.0  2.0
2  1.0  60.0  2.0
3  0.0  53.0  1.0
4  0.0  25.0  4.0

In [4]:
X1 = df['X1']
X2 = df['X2']
Y = df['Y']
n=len(X1)

### a. Find the maximum likelihood estimates of β0, β1, β2,. State the fitted responsefunction.

In [5]:
import statsmodels.api as sm

X = sm.add_constant(df[['X1', 'X2']])
print(X.head())
Y = df['Y']

logit_model = sm.GLM(Y, X, family=sm.families.Binomial())
logit_results = logit_model.fit()
print(logit_results.summary())

   const    X1   X2
0    1.0  32.0  3.0
1    1.0  45.0  2.0
2    1.0  60.0  2.0
3    1.0  53.0  1.0
4    1.0  25.0  4.0
                 Generalized Linear Model Regression Results                  
Dep. Variable:                      Y   No. Observations:                   33
Model:                            GLM   Df Residuals:                       30
Model Family:                Binomial   Df Model:                            2
Link Function:                  Logit   Scale:                          1.0000
Method:                          IRLS   Log-Likelihood:                -18.345
Date:                Wed, 11 Jan 2023   Deviance:                       36.690
Time:                        14:43:03   Pearson chi2:                     33.6
No. Iterations:                     4   Pseudo R-squ. (CS):             0.2223
Covariance Type:            nonrobust                                         
                 coef    std err          z      P>|z|      [0.025      0.975]
-----------

In [6]:
logit_results.params

const   -4.739309
X1       0.067733
X2       0.598632
dtype: float64

### Yi= -4.739309  + 0.067733X1 + 0.598632X2


### b. Obtain exp(b1), exp(b2). Interpret these numbers.

In [10]:
b1= 0.067733 
b2=  0.598632
betas=[b1,b2]
ORs=[]

In [11]:
for b in betas: 
    OR= np.exp(b)
    ORs.append(OR)
print(ORs)

[1.0700795590906513, 1.8196278460727975]


### Conclusion: 
- Y will increase by 1.07 times if X1 increases, as the exp(b1) = 1.075502506212986> 1
-Y will increase by 1.81 times if X2 increases, as the exp(b2) = 1.81> 1

### c. What is the estimated probability that X1=50, X2=3?

In [16]:
# a= -4.739309  + 0.067733X1 + 0.598632X2
# X1=50, X2= 3
#calculate the probability corresponding to X
def logistic(a):
  return (np.exp(a) / (1 + np.exp(a)))

a= -4.739309  + 0.067733*50 + 0.598632*3
pi = logistic(a)
print(pi)

0.609030072176333
